src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -267,16 +269,58 @@ def __init__(
             link = cache_entry.link
         ireq = make_install_req_from_link(link, template)
         assert ireq.link == link
-        if ireq.link.is_wheel and not ireq.link.is_file:
+        if ireq.link.is_wheel:
             wheel = Wheel(ireq.link.filename)
             wheel_name = canonicalize_name(wheel.name)
-            assert name == wheel_name, f"{name!r} != {wheel_name!r} for wheel"
+
+            # Validate the name
+            if not ireq.link.is_file:
+                assert name == wheel_name, f"{name!r} != {wheel_name!r} for wheel"
+            elif name != wheel_name:
+                deprecated(
+                    reason=(
+                        f"Got an name mismatch in wheel ({name} != {wheel_name}) "
+                        f"for {wheel.filename}, which is not permitted "
+                        "as per the wheel specification."
+                    ),
```

Comment:
We should also mention which one pip is going to use.

Reply:
We have name in there, along with wheel's name.

Reply:
Yes but we don’t say _which_ one of them we’re going to use as the package’s name in later parts of the installation.